### PR TITLE
Add interface for resource usage listener to spi.

### PR DESCRIPTION
**Pull request by toregge**
Propagate resource usage to service layer.

@geirst : please review
@vekterli : FYI

### CODE REVIEW

File: persistence/src/vespa/persistence/spi/resource_usage.h [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+namespace storage::spi {
+
+/*
+ * Class representing resource usage for persistence provider.
+ * Numbers are normalized to be between 0.0 and 1.0
+ */
+class ResourceUsage
```
Comment (geirst) [inline persistence/src/vespa/persistence/spi/resource_usage.h:9]: Consider making this class immutable, removing the setter functions.

File: searchcore/src/vespa/searchcore/proton/persistenceengine/resource_usage_tracker.h [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+#include <vector>
+
+namespace storage::spi { class IResourceUsageListener; }
+
+namespace proton {
+
+class DiskMemUsageState;
+class IDiskMemUsageNotifier;
+
+class ResourceUsageTracker : public IDiskMemUsageListener
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/persistenceengine/resource_usage_tracker.h:18]: Please add class description.

File: searchcore/src/vespa/searchcore/proton/server/proton.cpp [hunk 3]
```diff
@@ -310,8 +310,11 @@ Proton::init(const BootstrapConfig::SP & configSnapshot)
         protonConfig.basedir.c_str(), getcwd(tmp, sizeof(tmp)));
 
     _persistenceEngine = std::make_unique<PersistenceEngine>(*this, _diskMemUsageSampler->writeFilter(),
+                                                             _diskMemUsageSampler->notifier(),
                                                              protonConfig.visit.defaultserializedsize,
                                                              protonConfig.visit.ignoremaxbytes);
+    _diskMemUsageSampler->notifier().addDiskMemUsageListener(&_persistenceEngine->get_resource_usage_tracker());
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/server/proton.cpp:316]: Is this needed? It seems to already happen in the constructor of ResourceUsageTracker, which the PersistenceEngine instance owns.

File: searchcore/src/vespa/searchcore/proton/persistenceengine/resource_usage_tracker.h (hunk 2 above)
Comment (geirst) [reply searchcore/src/vespa/searchcore/proton/persistenceengine/resource_usage_tracker.h:18]: This class is currently not unit tested, but I guess that will be part of upcoming PRs.

**Review by geirst — APPROVED** [review]
👍

File: searchcore/src/tests/proton/persistenceengine/resource_usage_tracker/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (geirst) [inline searchcore/src/tests/proton/persistenceengine/resource_usage_tracker/CMakeLists.txt:1]: Please update copyright.

File: searchcore/src/vespa/searchcore/proton/test/disk_mem_usage_notifier.h [hunk 5]
```diff
@@ -19,7 +19,8 @@ class DiskMemUsageNotifier : public IDiskMemUsageNotifier
     std::vector<IDiskMemUsageListener *> _listeners;
     DiskMemUsageState _state;
 public:
-    DiskMemUsageNotifier() : IDiskMemUsageNotifier(), _listeners(), _state() { }
+    DiskMemUsageNotifier() : IDiskMemUsageNotifier(), _listeners(),
+                             _state({ 0.8, 0.5 }, { 0.8, 0.4 }) { }
```
Comment (geirst) [inline searchcore/src/vespa/searchcore/proton/test/disk_mem_usage_notifier.h:23]: Perhaps it is better to take these disk and memory parameters as constructor arguments in a new contructor? It was hard to understand why ResourceUsageTrackerTest::resource_usage_is_forwarded_to_listener() got the values 0.5 and 0.4 after just adding a listener to the tracker.

**Review by geirst — APPROVED** [review]
👍

File: persistence/src/vespa/persistence/spi/persistenceprovider.h [hunk 6]
```diff
@@ -374,6 +376,16 @@ struct PersistenceProvider
      * source1 and source2 should be stored in the target bucket.
      */
     virtual Result join(const Bucket& source1, const Bucket& source2, const Bucket& target, Context&) = 0;
+
+    /*
+     * Register a listener for updates to resource usage.
+     */
+    virtual void register_resource_usage_listener(std::shared_ptr<IResourceUsageListener> listener) = 0;
```
Comment (baldersheim) [inline persistence/src/vespa/persistence/spi/persistenceprovider.h:383]: Why do we need ways to register and unregister ?
Couldn't this just be an IResourceUsageListener & in the initialize method that is valid for the lifetime of the provider ?

Comment (toregge) [reply persistence/src/vespa/persistence/spi/persistenceprovider.h:383]: I'll remove the register / unregister methods and add argument to initialize as proposed.

**Pull request closed without merging**